From a4c31060cea725a2b0823916a2fb5249388fbd42 Mon Sep 17 00:00:00 2001
From: Igor Grinberg <igrinberg@habana.ai>
Date: Sun, 08 Aug 2021 20:31:30 +0300
Subject: [PATCH] [SW-52069] gaudi: security: protect sync mngr

Add Sync Manager protection using the protection bits.
Keep both of the registers open for access.
Effectively protecting the empty parts, internal configuration
registers, and the protection bits themselves.

Change-Id: Id44d58de9eb1c55494bd5ef2cbdce33441e7b4eb
Signed-off-by: Igor Grinberg <igrinberg@habana.ai>
---

diff --git a/arch/gaudi/arm/security.c b/arch/gaudi/arm/security.c
index 887cb52..576ed3c 100644
--- a/arch/gaudi/arm/security.c
+++ b/arch/gaudi/arm/security.c
@@ -602,6 +602,30 @@
 	}
 }
 
+#define SM_DIST		0x20000
+#define SM_ENGINES	4
+static void preboot_secure_sm(void)
+{
+	u32 prot_base, base, mask;
+	int i;
+
+	mask = ~(PROT_BIT(mmSYNC_MNGR_W_S_SYNC_MNGR_GLBL_SM_SEI_MASK) &
+		 PROT_BIT(mmSYNC_MNGR_W_S_SYNC_MNGR_GLBL_SM_SEI_CAUSE));
+
+	for (i = 0; i < SM_ENGINES; i++) {
+		base = (mmSYNC_MNGR_GLBL_W_S_BASE & ~HL_CFG_BASE) + i * SM_DIST;
+
+		/* offsets 0x0 - 0x80 */
+		prot_base = base;
+		regmap_writel(mask, PRIV_PROT_REG(prot_base));
+
+		/* offsets 0x80 - 0xF80 */
+		prot_base += SNGL_REG_PROT_SIZE;
+		set_prot_bits_range_all(prot_base,
+					PROT_SIZE_MAX - (prot_base - base));
+	}
+}
+
 static void preboot_secure_cpu(void)
 {
 	u32 prot_base, base = mmCPU_CA53_CFG_BASE & ~HL_CFG_BASE;
@@ -1826,6 +1850,7 @@
 		preboot_secure_if();
 		preboot_secure_sram_cfg();
 		preboot_secure_dma();
+		preboot_secure_sm();
 		preboot_secure_dbg();
 
 		preboot_secure_lbw();
